src/aspire/orientation/__init__.py

```diff
@@ -7,35 +15,434 @@ class Orient3D:
+
+        :param src: The source object of 2D denoised or class-averaged imag
+        :param nrad: The number of points in the radial direction
+        :param ntheta: The number of points in the theta direction
+        :param masked: Whether applying a mask to the images or not
         """
-        pass
+        self.src = src
+        self.nimg = src.n
+        self.nres = self.src.L
+        if nrad is None:
+            self.nrad = math.ceil(config.orient.r_ratio * self.nres)
+        else:
+            self.nrad = nrad
+        if ntheta is None:
+            self.ntheta = config.orient.n_theta
+        else:
+            self.ntheta = ntheta
+
+        self.basis = PolarBasis2D((self.nres, self.nres), self.nrad, self.ntheta)
+
+        self.max_shift = math.ceil(config.orient.max_shift * self.nres)
+        self.shift_step = config.orient.shift_step
+
+        imgs = self.src.images(start=0, num=np.inf).asnumpy()
 
-    def get_commonlines(self):
+        if masked:
+            mask_radius = self.nres * 0.45
+            self.rise_time = config.orient.rise_time
+            self.fuzzy_mask_dims = config.orient.fuzzy_mask_dims
+            # mask_radius is of the form xxx.5
+            if mask_radius * 2 == int(mask_radius * 2):
+                mask_radius = math.ceil(mask_radius)
+            # mask is not of the form xxx.5
+            else:
+                mask_radius = int(round(mask_radius))
+            # mask projections
+            m = self._fuzzy_mask(self.nres, mask_radius)
+            masked_projs = imgs.copy()
+            masked_projs = masked_projs.transpose((2, 0, 1))
+            masked_projs *= m
+            masked_projs = masked_projs.transpose((1, 2, 0))
+        else:
+            masked_projs = imgs
+
+        # Obtain coefficients in polar Fourier basis for input 2D images
+        self.basis = PolarBasis2D((self.nres, self.nres), self.nrad, self.ntheta)
+        self.pf = self.basis.evaluate_t(masked_projs)
+        self.pf = m_reshape(self.pf, (self.nrad, self.ntheta, self.nimg))
+
+        n_theta = self.ntheta
+        if n_theta % 2 == 1:
+            raise ValueError('n_theta must be even')
+        n_theta = n_theta // 2
+        self.pf = np.concatenate((np.flip(self.pf[1:, n_theta:], 0), self.pf[:, :n_theta]), 0)
+
+        self.clmatrix= None
+        self.cormatrix= None
+        self.shift_equations = None
+        self.shift_equations_map = None
+        self.clmatrix_mask = None
+        self.rotations = None
+
+    def estimate_rotations(self):
         """
-        find the common lines by voting method
+        Estimate orientation matrices for all 2D images
+
+        Subclasses should implement this function.
         """
-        pass
+        raise NotImplementedError('subclasses should implement this')
 
-    def build_clmatrix(self):
+    def output(self):
         """
-        build the common lines matrix
+        Output the 3D orientations in a star file
```

Comment:
In the new commit  0d1bfbc, the estimated rotation matrices are saved to a new ImageSource object so that they can be handled further by other modules.